pkg/ebpf/netlink/consumer.go

```diff
@@ -0,0 +1,290 @@
+			log.Debugf("rcv buffer size for netlink socket is %d bytes", size)
+		}
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+func (c *Consumer) Events() chan Event {
+	output := make(chan Event, outputBuffer)
+
+	c.do(false, func() {
+		c.conn.JoinGroup(netlinkCtNew)
+		c.receive(output, false)
+	})
+
+	return output
+}
+
+func (c *Consumer) DumpTable(family uint8) chan Event {
+	output := make(chan Event, outputBuffer)
+
+	c.do(false, func() {
+		defer close(output)
+
+		req := netlink.Message{
+			Header: netlink.Header{
+				Flags: netlink.Request | netlink.Dump,
+				Type:  netlink.HeaderType((unix.NFNL_SUBSYS_CTNETLINK << 8) | ipctnlMsgCtGet),
+			},
+			Data: []byte{family, unix.NFNETLINK_V0, 0, 0},
+		}
+
+		verify, err := c.conn.Send(req)
+		if err != nil {
+			output <- c.eventFor(nil, err)
+			return
+		}
+
+		if err := netlink.Validate(req, []netlink.Message{verify}); err != nil {
+			output <- c.eventFor(nil, err)
+			return
+		}
+
+		c.receive(output, true)
+	})
+
+	return output
+}
+
+func (c *Consumer) Stop() {
+	c.conn.Close()
+}
+
+func (c *Consumer) initWorker(procRoot string) {
+	go func() {
+		util.WithRootNS(procRoot, func() {
+			for {
+				fn, ok := <-c.workQueue
+				if !ok {
+					return
+				}
+				fn()
+			}
+		})
+	}()
+}
+
+func (c *Consumer) do(sync bool, fn func()) {
+	if !sync {
+		c.workQueue <- fn
+		return
+	}
+
+	done := make(chan struct{})
+	syncFn := func() {
+		fn()
+		close(done)
+	}
+	c.workQueue <- syncFn
+	<-done
+}
+
+// Based on of https://github.com/mdlayher/netlink/conn.go
+// In our context we don't care about multi-part messages in the sense
+// that is better to flush their parts to the output channel (and generate back-pressure)
+// instead of buffering the whole thing in memory. When dump is set to true we terminate
+// after processing the first multi-part message.
+func (c *Consumer) receive(output chan Event, dump bool) {
+	for {
+		c.pool.Reset()
+		msgs, err := c.socket.Receive()
+		if err != nil {
+			if !isEOF(err) {
+				output <- c.eventFor(nil, err)
+			}
+			return
+		}
+
+		for _, m := range msgs {
+			if err := checkMessage(m); err != nil {
+				output <- c.eventFor(nil, err)
+				return
+			}
+		}
+
+		multiPartDone := len(msgs) > 0 && msgs[len(msgs)-1].Header.Type == netlink.Done
+		if multiPartDone {
+			msgs = msgs[:len(msgs)-1]
+		}
+
+		output <- c.eventFor(msgs, nil)
+
+		// If we're doing a conntrack dump it means we are done after reading the multi-part message
+		if dump && multiPartDone {
+			return
+		}
+	}
+}
+
+func (c *Consumer) eventFor(msgs []netlink.Message, err error) Event {
+	return Event{
+		msgs:    msgs,
+		err:     err,
+		buffers: c.pool.inUse,
+		pool:    c.pool,
+	}
+}
+
+type bufferPool struct {
+	inUse [][]byte
+	sync.Pool
+}
+
+func newBufferPool() *bufferPool {
+	return &bufferPool{
+		Pool: sync.Pool{
+			New: func() interface{} {
+				return make([]byte, os.Getpagesize())
```

Comment:
Can you elaborate? Is the slice descriptor also allocated in the heap?

Reply:
From the [documentation](https://golang.org/pkg/sync/#example_Pool):

> 	New: func() interface{} {
		// The Pool's New function should generally only return pointer
		// types, since a pointer can be put into the return interface
		// value without an allocation:
		return new(bytes.Buffer)
	},

This benchmark illustrates it

```
func BenchmarkSlicePool(b *testing.B) {
	b.Run("nopointer", func(b *testing.B) {
		pool := sync.Pool{
			New: func() interface{} {
				return make([]byte, 10)
			},
		}
		b.ReportAllocs()

		for i := 0; i < b.N; i++ {
			buf := pool.Get().([]byte)
			buf[0] = 'x'
			pool.Put(buf)
		}
	})

	b.Run("pointer", func(b *testing.B) {
		pool := sync.Pool{
			New: func() interface{} {
				slice := make([]byte, 10)
				return &slice
			},
		}
		b.ReportAllocs()

		for i := 0; i < b.N; i++ {
			x := pool.Get().(*[]byte)
			buf := *x
			buf[0] = 'x'
			pool.Put(x)
		}
	})
}
```

```
BenchmarkSlicePool/nopointer-2         	20062760	        60.4 ns/op	      32 B/op	       1 allocs/op
BenchmarkSlicePool/pointer-2           	73635993	        15.4 ns/op	       0 B/op	       0 allocs/op
```

Reply:
Cool, though I don't understand it. 😕 Would mind you mind explaining this?
> since a pointer can be put into the return interface value without an allocation

Reply:
I believe that you were on the right track with

> Is the slice descriptor also allocated in the heap?

The slice descriptor is itself a value and when you return that as an interface, it gets heap allocated.  Returning pointers as interfaces does not require that conversation

Reply:
Oh, I see, so it's because of the interface. Cool, thanks for clarifying :)